crates/wasmparser/src/readers/core/linking.rs

```diff
@@ -0,0 +1,405 @@
+    InitFuncs(InitFuncMap<'a>),
+    /// The [COMDAT](https://llvm.org/docs/LangRef.html#comdats) groups of associated linking objects.
+    ComdatInfo(ComdatMap<'a>),
+    /// Extra information about the symbols present in the module.
+    SymbolTable(SymbolInfoMap<'a>),
+    /// An unknown [linking subsection](https://github.com/WebAssembly/tool-conventions/blob/main/Linking.md#linking-metadata-section).
+    Unknown {
+        /// The identifier for this subsection.
+        ty: u8,
+        /// The contents of this subsection.
+        data: &'a [u8],
+        /// The range of bytes, relative to the start of the original data
+        /// stream, that the contents of this subsection reside in.
+        range: Range<usize>,
+    },
+}
+
+impl<'a> Subsection<'a> for Linking<'a> {
+    fn from_reader(id: u8, reader: BinaryReader<'a>) -> Result<Self> {
+        let data = reader.remaining_buffer();
+        let offset = reader.original_position();
+        Ok(match id {
+            5 => Self::SegmentInfo(SegmentMap::new(data, offset)?),
+            6 => Self::InitFuncs(InitFuncMap::new(data, offset)?),
+            7 => Self::ComdatInfo(ComdatMap::new(data, offset)?),
+            8 => Self::SymbolTable(SymbolInfoMap::new(data, offset)?),
+            ty => Self::Unknown {
+                ty,
+                data,
+                range: offset..offset + data.len(),
+            },
+        })
+    }
+}
+
+impl<'a> LinkingSectionReader<'a> {
+    /// Creates a new reader for the linking section contents starting at
+    /// `offset` within the original wasm file.
+    pub fn new(data: &'a [u8], offset: usize) -> Result<Self> {
+        let mut reader = BinaryReader::new_with_offset(data, offset);
+        let range = reader.range();
+
+        let version = reader.read_var_u32()?;
+        if version != 2 {
+            return Err(BinaryReaderError::new(
+                format!("unsupported linking section version: {}", version),
+                offset,
+            ));
+        }
+
+        let subsections = Subsections::new(reader.remaining_buffer(), reader.original_position());
+        Ok(Self {
+            version,
+            subsections,
+            range,
+        })
+    }
+
+    /// Returns the version of linking metadata contained in this section.
+    pub fn version(&self) -> u32 {
+        self.version
+    }
+
+    /// Returns the original byte offset of this section.
+    pub fn original_position(&self) -> usize {
+        self.subsections.original_position()
+    }
+
+    /// Returns the range, as byte offsets, of this section within the original
+    /// wasm binary.
+    pub fn range(&self) -> Range<usize> {
+        self.range.clone()
+    }
+
+    /// Returns the iterator for advancing through the subsections.
+    ///
+    /// You can also use [`IntoIterator::into_iter`] directly on this type.
+    pub fn subsections(&mut self) -> &mut Subsections<'a, Linking<'a>> {
+        &mut self.subsections
+    }
```

Comment:
Could this return a fresh copy of the iterator rather than a mutable view into the internals? Idiomatically most of `wasmparser` tries to avoid internal mutation and return sub-view except for `Iterator`-like things on `next` methods.